Prepare IntervalTask for checkpoint

tjwatson opened the pull request:
Block work from IntervalTask once we have prepared for checkpoint This is to avoid deadlocks that can be possible once we go into single-thread mode for a checkpoint


- [ ] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [ ] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [ ] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

tjwatson committed:
Prepare IntervalTask for checkpoint

Block work from IntervalTask once we have prepared for checkpoint
This is to avoid deadlocks that can be possible once we go into
single-thread mode for a checkpoint

tjwatson commented:
#build

SmithaSubbarao reviewed (dismissed):
/lgtm

Based on the explanation from the [defect 302615](https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.workitem.viewWorkItem&id=302615): 
```

The javacore produced show these two threads.  The first one is the single-thread running during checkpoint/restore.  The second one is a thread that was running on the checkpoint side before entering single-thread mode. The IntervalTask should probably protect against entering com.ibm.ws.threading.internal.ThreadPoolController.evaluateInterval() after preparing for a checkpoint.
```

tjwatson commented:
#build

tjwatson commented:
#build

gjdeval commented on dev/com.ibm.ws.kernel.boot.common/src/io/openliberty/checkpoint/spi/CheckpointPhase.java:
```diff
@@ -125,6 +126,35 @@ static synchronized void setPhase(String p) {
+            phase.addMultiThreadedHook(new CheckpointHook() {
+                @Override
+                public void prepare() {
+                    debug(() -> "Locking locking checkpoint write lock for prepare");
```
"Locking locking"?

tjwatson replied:
> "Locking locking"?

Should be "Locking checkpoint write lock for prepare"

tjwatson replied:
Fixed.

tjwatson commented on dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java:
```diff
@@ -2197,8 +2192,8 @@ synchronized void introspect(PrintWriter out) {
  */
-class IntervalTask extends TimerTask {
-
+class IntervalTask extends TimerTask implements CheckpointHook {
```
I need to fix this up.  Should not implement `CheckpointHook`

tjwatson replied:
Fixed

tjwatson commented on dev/com.ibm.ws.kernel.boot.common/src/io/openliberty/checkpoint/spi/CheckpointPhase.java:
```diff
@@ -125,6 +126,35 @@ static synchronized void setPhase(String p) {
         }
         THE_PHASE = phase;
+        if (phase != INACTIVE) {
+            phase.addMultiThreadedHook(new CheckpointHook() {
```
I should add a rank here so we obtain the write lock as late as possible before going into single-thread mode.

tjwatson replied:
fixed

gjdeval commented on dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java:
```diff
@@ -2208,7 +2203,7 @@ class IntervalTask extends TimerTask {
     public void run() {
         try {
-            threadPoolController.evaluateInterval();
+            phase.runWithCheckpointLock(threadPoolController::evaluateInterval);
```
I do not know anything about how checkpoint/restore works at the implementation level, so I cannot helpfully review that part of the changes. :-) 

It looks like `runWithCheckpointLock` will be very low overhead the vast majority of the time when checkpoint/restore functions are not busy doing some work - is that correct? 

If so, then I don't see any problem with this change. The controller cycle runs every 1500 ms (if coreThreads != maxThreads) so we would not want to add any significant burden to that operation.

tjwatson replied:
> It looks like `runWithCheckpointLock` will be very low overhead the vast majority of the time when checkpoint/restore functions are not busy doing some work - is that correct?

Correct, it should be a no-op for the "normal" non-checkpoint case and once we have restored for the checkpoint case.

tjwatson commented:
#build

tjwatson commented:
#build

gjdeval reviewed (dismissed):
I do not see any problems with the ThreadPoolController interaction with checkpoint.

tjwatson committed:
Write debug message after obtaining lock

Avoids out of order locks if some other thread is trying
to write to the console while holding the read checkpoint lock.

tjwatson committed:
Obtain the write lock for logging in checkpoint as late as possible

Need to obtain the write lock as late as possible to avoid contention
on the read lock and out of order locks with the checkpoint lock

tjwatson committed:
Tests for runWithCheckpointLock

tjwatson commented:
#build

dazavala reviewed (approved):
Looks good.  My only concern is whether the call to acquire (lock) the writeLock in CheckpointPhase,prepare() might fail because a thread running ThreadPoolController.evaluateInterval() has acquired the readLock() at the same time.

> RRWL.WriteLock.lock():  Acquires the write lock if neither the read nor write lock are held by another thread and returns immediately, setting the write lock hold count to one.

tjwatson commented:
> Looks good. My only concern is whether the call to acquire (lock) the writeLock in CheckpointPhae,prepare() might fail because a thread running ThreadPoolController.evaluateInterval() has acquired the readLock() at the same time.
> 
> > RRWL.WriteLock.lock():  Acquires the write lock if neither the read nor write lock are held by another thread and returns immediately, setting the write lock hold count to one.

That would be possible if the function passed to `io.openliberty.checkpoint.spi.CheckpointPhase.runWithCheckpointLock(WithCheckpointLock<R, T>)` then became the thread to triggered a checkpoint to happen such that the thread used to perform the checkpoint would already have the read-lock while calling the function.  That is not possible for the current, single usage of the `runWithCheckpointLock`  because that function doesn't do anything that would cause the thread to trigger a checkpoint.

I think any usage of the method with a function that triggers a checkpoint should be invalid and I think it is fine for it to throw an exception for now.  If this becomes a problem then we may need to do some special handling where the prepare gives up the read lock(s) in order to obtain the write lock and then re-acquires the read lock(s) while holding the write lock before continuing.
